src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs

```diff
@@ -216,25 +216,36 @@ public CopyPosition CopyTo(CopyPosition position, T[] array, int arrayIndex, int
             int row = position.Row;
             int column = position.Column;
 
-            for (; count > 0; row++, column = 0)
+            T[] buffer = GetBuffer(row);
+            int copied = CopyToCore(buffer, column);
+
+            while (count > 0)
+            {
+                buffer = GetBuffer(++row);
+                copied = CopyToCore(buffer, 0);
+            }
+
+            return copied == buffer.Length
+                ? new CopyPosition(row + 1, 0)
+                : new CopyPosition(row, copied);
+
+            int CopyToCore(T[] sourceBuffer, int sourceIndex)
             {
-                T[] buffer = GetBuffer(index: row);
+                Debug.Assert(sourceBuffer.Length > sourceIndex);
 
-                // During this iteration, copy until we satisfy `count` or reach the
-                // end of the current buffer.
-                int copyCount = Math.Min(buffer.Length, count);
+                // Copy until we satisfy `count` or reach the end of the current buffer.
+                int copyCount = Math.Min(sourceBuffer.Length - sourceIndex, count);
 
                 if (copyCount > 0)
```

Comment:
We don't have any tests where copyCount == 0.

Reply:
@OmarTawfik I don't think this can happen anymore. Add these asserts and remove the conditional:

```diff
+Debug.Assert(count > 0);
 Debug.Assert(sourceBuffer.Length > sourceIndex);

 // Copy until we satisfy `count` or reach the end of the current buffer.
 int copyCount = Math.Min(sourceBuffer.Length - sourceIndex, count);
+Debug.Assert(copyCount > 0);
```